### ISSUE TITLE
Truncated Backpropagation Through Time

### ISSUE THREAD

**Issue by Rikorose** [williamFalcon/pytorch-lightning#293]
**Is your feature request related to a problem? Please describe.**
Training very long sequences using RNNs is not easily feasible. Thus, limiting the number of steps for forward and backward path, while keeping the last state for the next iterateion, speeds up training significantly.

**Describe the solution you'd like**
Have an option in trainer like `bptt_steps`

**Describe alternatives you've considered**
Overwrite the trainers ´__run_training_batch´ method.

**Comment by Rikorose** [comment]
Here is an implementation for 2 variable length sequences. Input is a packed sequence, so that the zero padding does not need copied to GPU, padding can be done after copying.

```pyimport torch
import torch.nn as nn

n_features = 1
out_size = 2
tbptt_steps = 3

seq_len_1 = 3
seq_len_2 = 5

batch_first = True
split_dim = 1 if batch_first else 0

x = [torch.rand(seq_len_1, n_features), torch.rand(seq_len_2, n_features)]
y = [torch.rand(seq_len_1, out_size), torch.rand(seq_len_2, out_size)]
x = nn.utils.rnn.pack_sequence(x, enforce_sorted=False)
y = nn.utils.rnn.pack_sequence(y, enforce_sorted=False)

model = nn.RNN(
    input_size=n_features, hidden_size=out_size, num_layers=2, batch_first=batch_first
)

opt = torch.optim.Adam(model.parameters())

# In lightning module
x, x_lengths = nn.utils.rnn.pad_packed_sequence(x, batch_first=batch_first)
y, y_lengths = nn.utils.rnn.pad_packed_sequence(y, batch_first=batch_first)

for i, (x_, y_) in enumerate(
    zip(x.split(tbptt_steps, dim=split_dim), y.split(tbptt_steps, dim=split_dim))
):
    print(f"{i}: {x_.shape}")
    # Forward path
    y_pred, state = model(x_)
    # Compute loss
    loss = nn.functional.mse_loss(y_, y_pred)
    # Backward path
    loss.backward()
    # Update weights
    opt.step()
```

Output with `batch_first = True`:
```
0: torch.Size([2, 3, 1])
1: torch.Size([2, 2, 1])
```

Output with `batch_first = False`:
```
0: torch.Size([3, 2, 1])
1: torch.Size([2, 2, 1])
```

**Comment by Rikorose** [comment]
Yes this sounds good.

The state however is not handled in the current trainer right now.
- The state needs to be passed from a tbptt step to the next
- Before providing it to the model, it has to be detached

I think the following should be done in the LightningModule and not in the dataloader due to several reasons:
```py
x, y = batch
if input_is_packed:
    x, x_lengths = nn.utils.rnn.pad_packed_sequence(x, batch_first=batch_first)
    y, y_lengths = nn.utils.rnn.pad_packed_sequence(y, batch_first=batch_first)
```
1. Only the meaningful data and no zeros of the padding need to be copied
2. `batch` might not be a tuple but a custom type, e.g. 
```py
batch = {
  'input': input_tensor,
  'target': target_tensor,
  'meta': some_meta_dict,
}
```
So we cannot do ´x, y = batch´.

Both can be solved if `split_each_element_in_batch_fx` is implemented in the LightningModule. Here the user can pad, process the custom input batch and split the data. We could also provide a default implementation like above via `x.split(tbptt_steps)`.
